OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java

```diff
@@ -572,139 +443,177 @@ private static OSEmailSubscriptionState getLastEmailSubscriptionState(Context co
+      notificationWillShowInForegroundHandler = callback;
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId, NotificationOpenedHandler notificationOpenedHandler) {
-      init(context, googleProjectNumber, oneSignalAppId, notificationOpenedHandler, null);
+   public static void setNotificationOpenedHandler(NotificationOpenedHandler callback) {
+      notificationOpenedHandler = callback;
+      if (appContext != null)
+        setAppContext(appContext);
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId, NotificationOpenedHandler notificationOpenedHandler, NotificationReceivedHandler notificationReceivedHandler) {
-      mInitBuilder = createInitBuilder(notificationOpenedHandler, notificationReceivedHandler);
-      OneSignal.setAppContext(context);
-      setupPrivacyConsent(context);
+   public static void setInAppMessageClickHandler(InAppMessageClickHandler callback) {
+      inAppMessageClickHandler = callback;
+   }
 
+   /**
+    * Called after setAppId and setAppContext, depending on which one is called last (order does not matter)
+    */
+   synchronized private static void init() {
       if (requiresUserPrivacyConsent()) {
-         OneSignal.Log(LOG_LEVEL.VERBOSE, "OneSignal SDK initialization delayed, user privacy consent is set to required for this application.");
-         delayedInitParams = new DelayedConsentInitializationParameters(context, googleProjectNumber, oneSignalAppId, notificationOpenedHandler, notificationReceivedHandler);
+         OneSignal.Log(LOG_LEVEL.WARN, "OneSignal SDK initialization delayed, user privacy consent is set to required for this application.");
+         delayedInitParams = new DelayedConsentInitializationParameters(appContext, googleProjectNumber, appId, notificationWillShowInForegroundHandler, notificationOpenedHandler);
+         // Set app id null since OneSignal was not init fully
+         appId = null;
          return;
       }
 
-      mInitBuilder = createInitBuilder(notificationOpenedHandler, notificationReceivedHandler);
-
       if (!isGoogleProjectNumberRemote())
-         mGoogleProjectNumber = googleProjectNumber;
+         googleProjectNumber = "";
 
       deviceType = osUtils.getDeviceType();
-      subscribableStatus = osUtils.initializationChecker(context, deviceType, oneSignalAppId);
+      subscribableStatus = osUtils.initializationChecker(appContext, deviceType, appId);
       if (isSubscriptionStatusUninitializable())
          return;
 
-      // Pre-check on app id to make sure init of SDK is performed properly
-      //    Usually when the app id is changed during runtime so that SDK is reinitialized properly
-      if (appId != null && !appId.equals(oneSignalAppId))
-         initDone = false;
-
       if (initDone) {
-         if (mInitBuilder.mNotificationOpenedHandler != null)
+         if (notificationOpenedHandler != null)
```

Comment:
I think we can clean this up since we don't have an external init that someone can call more than once.